### PR TITLE
Implement git-machete 2.14.0 compatible discover on backend

### PR DESCRIPTION
**Pull request by PawelLipski**

### CODE REVIEW

File: testCommon/src/test/resources/common.sh [hunk 1]
```diff
@@ -2,7 +2,9 @@
 
 status_branch_hook=$(cat <<'EOF'
 #!/usr/bin/env bash
-git log -1 --format=%cd  # commit timestamp of the branch tip
+branch=$1
+file_count=$(git ls-tree $branch | wc -l)
```
Comment (PawelLipski) [inline testCommon/src/test/resources/common.sh:6]: Unrelate change to .git/hooks/machete-status-branch scrip used in the tests.

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java [hunk 2]
```diff
@@ -629,50 +695,87 @@ private SyncToParentStatus deriveSyncToParentStatus(
...
+        subentries = subentries.append(newSubentry);
+      }
 
-      if (commitAndContainingBranches != null) {
-        var commit = commitAndContainingBranches._1;
-        var containingBranches = commitAndContainingBranches._2.toList();
-        assert containingBranches.nonEmpty() : "containingBranches is empty";
+      MyBranchLayoutEntry getRoot() {
+        // Path compression to reduce the lookup time,
+        // see https://en.wikipedia.org/wiki/Disjoint-set_data_structure#Find
```
Comment (mkondratek) [inline backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java:736]: :+1:

File: backendImpl/src/test/java/com/virtuslab/gitmachete/backend/integration/GitMacheteStatusTestSuite.java [hunk 3]
```diff
@@ -55,26 +56,45 @@
...
-  @SneakyThrows
   public GitMacheteStatusTestSuite(String scriptName) {
     super(scriptName);
-    IBranchLayout branchLayout = branchLayoutReader.read(repositoryGitDir.resolve("machete"));
-    gitMacheteRepository = gitMacheteRepositoryFactory.create(repositoryMainDir, repositoryGitDir, branchLayout);
   }
 
-  @Test
-  public void hasSameOutputAsCli() {
+  //@Test
```
Comment (mkondratek) [inline backendImpl/src/test/java/com/virtuslab/gitmachete/backend/integration/GitMacheteStatusTestSuite.java:63]: :confused: :question:

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java [hunk 4]
```diff
@@ -629,50 +695,87 @@ private SyncToParentStatus deriveSyncToParentStatus(
...
-        return Option.some(firstContainingBranchName);
-      } else {
-        LOG.debug(() -> "Could not infer upstream for branch '${branch.getFullName()}'");
+      @Override
+      public Option<String> getCustomAnnotation() {
         return Option.none();
       }
     }
+
+    IGitMacheteRepository discoverGitMacheteRepository() throws GitMacheteException, GitCoreException {
```
Comment (mkondratek) [inline backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java:749]: What about wrapping `GitCoreException` into `GitMacheteException` to declare only the higher-level one?

File: backendImpl/src/test/java/com/virtuslab/gitmachete/backend/integration/GitMacheteStatusTestSuite.java (hunk 3 above)
Comment (PawelLipski) [reply backendImpl/src/test/java/com/virtuslab/gitmachete/backend/integration/GitMacheteStatusTestSuite.java:63]: Whoops... fixed

File: backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java (hunk 4 above)
Comment (PawelLipski) [reply backendImpl/src/main/java/com/virtuslab/gitmachete/backend/impl/GitMacheteRepositoryFactory.java:749]: Done

File: backendImpl/src/test/java/com/virtuslab/gitmachete/backend/integration/GitMacheteStatusTestSuite.java [hunk 5]
```diff
@@ -38,12 +41,31 @@
...
+    process.waitFor(1, TimeUnit.SECONDS);
+    var exitValue = process.exitValue();
+    if (exitValue != 0) {
+      Assert.fail("git-machete CLI is not installed");
+    }
+    var version = new String(process.getInputStream().readAllBytes())
+        .stripTrailing()
+        .replace("git-machete version ", "");
+    if (!version.equals(CLI_REFERENCE_VERSION)) {
+      Assert.fail("git-machete is expected in version ${CLI_REFERENCE_VERSION}, found ${version}");
```
Comment (mkondratek) [inline backendImpl/src/test/java/com/virtuslab/gitmachete/backend/integration/GitMacheteStatusTestSuite.java:65]: :+1:

File: scripts/run-pre-build-checks [hunk 6]
```diff
@@ -19,6 +19,7 @@ bash ${x_option:+-x} ./scripts/enforce-version-bump
 
 ./scripts/prohibit-class-name-hardcode-in-strings
 ./scripts/prohibit-inner-classes
+./scripts/prohibit-java-code-commented-out
```
Comment (mkondratek) [inline scripts/run-pre-build-checks:22]: :care:

**Pull request merged**